File: day3/cxx/puzzle1.cxx
#include <algorithm>
#include <fstream>
#include <iostream>
#include <set>
#include <string>
#include <vector>

#include "lines.hpp"
#include "util.hpp"

using namespace std;

const auto solve()
{
    ifstream f{"../input.txt"};

    auto sum_priorities = 0u;

    for (auto & line: lines(f))
    {
        string items = line;

        auto num_items = items.length() / 2;

        set<char> comp_1_set{items.begin(), items.begin() + num_items};
        set<char> comp_2_set{items.begin() + num_items, items.end()};

        vector<char> common(items.length());

        set_intersection(
            comp_1_set.begin(), comp_1_set.end(),
            comp_2_set.begin(), comp_2_set.end(),
            common.begin());

        sum_priorities += get_priority(common[0]);
    }

    return sum_priorities;
}

int main()
{
    auto solution = solve();
    cout << solution << endl;
}

File: day3/py/puzzle1.py
#!/usr/bin/env python

import itertools

from util import get_priority

def solve():

    with open('../input.txt', 'r') as f:
        sum_priorities = 0

        for line in f:
            line = line.strip() # remove newline, if present

            num_items = int(len(line)/2)
            
            comp_1 = itertools.islice(line, 0, num_items)
            comp_2 = itertools.islice(line, num_items, None)

            common = set(comp_1).intersection(set(comp_2)).pop()

            sum_priorities += get_priority(common)

    return sum_priorities

if __name__ == "__main__":
    solution = solve()
    print(solution, )

